### PR TITLE
feat: use testcontainers-go for clickhouse tests

### PR DESCRIPTION
**Pull request by mdelapenya**
Hello from the [Testcontainers for Go](https://github.com/testcontainers/testcontainers-go/) community :wave:!

This PR is a PoC for adopting testcontainers-go as the testing library for integration testing of the different storage integrations living in this project, for which we do have [Go modules](https://testcontainers.com/modules/?language=go) representing the majority of them. Please let me know if you have any concern about adding it, I'm really happy to assist in addressing them. Of course I'd understand any result for the PR.

## What does this PR do?

This PR adds testcontainers-go to the Clickhouse module, updating the function that expects a local instance running on 127.0.0.1:9001 to be present on the system. Instead, testcontainers-go will start the instance from the test sourcecode.

## Why is it important?
Less setup, as developers will just clone the repository and run the tests (given Docker or another container runtime 100% compatible is installed in the machine). Therefore, the CI experience is exactly the same as running the tests locally, which improves the debuggability.

As a side note, GH actions has Docker already installed, so it should work transparently.

## How to test this PR?

Just go and do like the CI does:

```shell
cd ./clickhouse && go clean -testcache && go test ./... -v -race
```
## Follow-up

If considered positively, add more PRs for the different storages.

## Summary by CodeRabbit

- **Documentation**
	- Enhanced the `README.md` with new sections on testing and local development, including requirements for Docker and usage of Testcontainers for integration tests.
	- Restructured the "Local development" section for improved clarity on setting up a Clickhouse cluster.

- **Tests**
	- Improved the ClickHouse testing suite by integrating dynamic configuration for connection settings using Testcontainers, enhancing test reliability and flexibility.
	- Streamlined the GitHub Actions workflows by removing direct Clickhouse startup steps and introducing environment variables for image reference, simplifying the testing process.

### CODE REVIEW

File: clickhouse/README.md [hunk 1]
```diff
@@ -31,7 +31,13 @@ Install the clickhouse implementation:
 go get github.com/gofiber/storage/clickhouse
 ```
 
-Before running or testing this implementation, you must ensure a Clickhouse cluster is available.
+### Running the tests
+
+This module uses Testcontainers for Go to run integration tests, which will start a local instance of Clickhouse as a Docker container under the hood. To run the tests, you must have Docker installed on your machine.
```
Comment (mdelapenya) [inline clickhouse/README.md:36]:
```suggestion
This module uses [Testcontainers for Go](https://github.com/testcontainers/testcontainers-go/) to run integration tests, which will start a local instance of Clickhouse as a Docker container under the hood. To run the tests, you must have Docker installed on your machine.
```

**Comment by gaby** [comment]
@mdelapenya Thank you for this, I'm a huge fan of TestContainers.

In the workflow yaml for Clickhouse you need to remove the ClickHouse instance we are starting else there will be two running.

Is there a way to get the clickhouse version from ENV? So that it's not hardcoded, or maybe use a major semver docker mage tag.

**Comment by mdelapenya** [comment]
> @mdelapenya Thank you for this, I'm a huge fan of TestContainers.
> 
> In the workflow yaml for Clickhouse you need to remove the ClickHouse instance we are starting else there will be two running.

Ah good catch! I'm fixing that immediately.

> Is there a way to get the clickhouse version from ENV? So that it's not hardcoded, or maybe use a major semver docker mage tag.

Indeed, it's totally possible, although I see more work to do in the benchmarks workflow. Is there a reason to not include it for each storage file. My concern is about defining lots of env vars for each storage in the same step, instead of having them self-contained in the test code. Here:

```yaml
- name: Run Benchmarks
        run: |
          set -o pipefail
          for d in */ ; do
            [[ $d == "tls/" ]] && continue
            [[ $d == "node_modules/" ]] && continue

            cd "$d"
            echo "Bench dir: $d"
            go test ./... -benchmem -run=^$ -bench . | tee -a ../output.txt
            cd ..
          done
        shell: bash
        env:
          MSSQL_DATABASE: master
          MSSQL_USERNAME: sa
          MSSQL_PASSWORD: MsSql!1234
          MYSQL_USERNAME: username
          MYSQL_PASSWORD: password
          MYSQL_DATABASE: fiber
          POSTGRES_DATABASE: fiber
          POSTGRES_USERNAME: username
          POSTGRES_PASSWORD: "pass#w%rd"
```

Do you think it's possible to move them to each storage's file? Will it affect the existing github checks, and their usages?

**Comment by ReneWerner87** [comment]
@gaby can you re-review

**Review by gaby — APPROVED** [review]
👍 LGTM

**Comment by mdelapenya** [comment]
I do not see that error while running the benchmarks in my machine (MacOS M1). Any clue?

**Comment by gaby** [comment]
@mdelapenya We were just talking about that internally. The main `go.mod` file uses `go1.20`, but clickhouse requires `go1.21` minimum.

The `go.mod` in `storage/clickhouse` uses `1.21`.

**Comment by gaby** [comment]
@mdelapenya @ReneWerner87 CI is passing now.

**Comment by mdelapenya** [comment]
Cool, thanks for merging this!

Do you have plan for adding testcontainers to the rest of the storages? I can create a tracking issue with the rest of them, so whoever wants to create a PR just needs to be assigned to it. Wdyt?

**Comment by gaby** [comment]
> Cool, thanks for merging this!
> 
> Do you have plan for adding testcontainers to the rest of the storages? I can create a tracking issue with the rest of them, so whoever wants to create a PR just needs to be assigned to it. Wdyt?

Yes, go ahead and create the issue. Thank you! 💪

**Comment by mdelapenya** [comment]
There it goes! https://github.com/gofiber/storage/issues/1507

**Pull request merged**